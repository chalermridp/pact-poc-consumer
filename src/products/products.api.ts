import { Injectable } from '@nestjs/common';
import axios from 'axios';

@Injectable()
export class ProductsAPI {
  private url = 'http://localhost:3000';

  setUrl(url: string) {
    this.url = url;
  }

  async getAll() {
    return axios.get(`${this.url}/products`);
  }

  async getById(id) {
    return axios.get(`${this.url}/products/${id}`);
  }
}
