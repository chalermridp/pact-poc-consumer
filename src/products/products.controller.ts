import { Controller, Get, Param } from '@nestjs/common';
import { BaseResponse } from '../common/responses/base.reponse';
import { ProductsService } from './products.service';
import { ProductResponse } from './responses/product.response';
import { ProductsResponse } from './responses/products.response';

@Controller('products')
export class ProductsController {
  constructor(private productsService: ProductsService) {}

  @Get()
  async getAllProducts(): Promise<BaseResponse<ProductsResponse>> {
    return new BaseResponse(
      200,
      new ProductsResponse(await this.productsService.getAll()),
    );
  }

  @Get('/:id')
  async getById(
    @Param('id') id: string,
  ): Promise<BaseResponse<ProductResponse>> {
    return new BaseResponse(
      200,
      new ProductResponse(await this.productsService.getById(id)),
    );
  }
}
