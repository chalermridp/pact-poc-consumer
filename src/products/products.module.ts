import { Module } from '@nestjs/common';
import { ProductsService } from './products.service';
import { ProductsController } from './products.controller';
import { ConfigModule } from '@nestjs/config';

@Module({
  imports: [ConfigModule],
  providers: [ProductsService],
  controllers: [ProductsController],
})
export class ProductsModule {}
