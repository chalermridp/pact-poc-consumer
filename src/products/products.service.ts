import { Injectable } from '@nestjs/common';
import { BaseException } from '../common/exceptions/base.exception';
import { ProductsAPI } from './products.api';
import { ProductsModel } from './products.model';

@Injectable()
export class ProductsService {
  constructor(private productsApi: ProductsAPI) { }

  async getAll(): Promise<ProductsModel[]> {
    const response = await this.productsApi.getAll();
    return response.data.data.products;
  }

  async getById(id: string): Promise<ProductsModel> {
    try {
      const response = await this.productsApi.getById(id);
      return response.data.data.product;
    } catch (error) {
      throw new BaseException(
        error.response.data.data.code,
        error.response.data.data.error_name,
        error.response.data.data.error_message,
      );
    }
  }
}
