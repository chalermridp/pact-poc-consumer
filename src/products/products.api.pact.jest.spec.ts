import { pactWith } from 'jest-pact';
import { Test } from '@nestjs/testing/test';
import { ProductsAPI } from './products.api';
import { ProductsModule } from './products.module';
import { like } from '@pact-foundation/pact/dsl/matchers';

pactWith(
  { consumer: 'PactPocConsumer', provider: 'PactPocProvider' },
  (provider) => {
    const allProductsResponse = {
      code: 200,
      data: {
        products: [
          {
            id: '1',
            sku: '111',
            name_en: 'water',
            price: 30,
            is_active: true,
          },
          {
            id: '2',
            sku: '222',
            name_en: 'shoes',
            price: 40,
            is_active: false,
          },
        ],
      },
    };

    const productResponse = {
      code: 200,
      data: {
        product: {
          id: '1',
          sku: '111',
          name_en: 'water',
          price: 30,
          is_active: true,
        },
      },
    };

    const productNotFoundErrorResponse = {
      code: 404,
      error_name: 'not_found',
      error_message: 'product not found',
    };

    let productsApi;
    beforeAll(async () => {
      const moduleRef = await Test.createTestingModule({
        imports: [ProductsModule],
      }).compile();

      productsApi = moduleRef.get(ProductsAPI);
      // productsApi.setUrl(provider.mockService.baseUrl);
      productsApi.setUrl('http://localhost:3000');
    });

    describe('get all products', () => {
      // beforeAll(() => {
      //   provider.addInteraction({
      //     state: 'products exist',
      //     uponReceiving: 'get all products',
      //     withRequest: {
      //       method: 'GET',
      //       path: '/products',
      //     },
      //     willRespondWith: {
      //       status: 200,
      //       headers: {
      //         'Content-Type': 'application/json; charset=utf-8',
      //       },
      //       body: like(allProductsResponse),
      //     },
      //   });
      // });

      it('return all products', (done) => {
        productsApi.getAll().then((response) => {
          expect(response.status).toBe(200);
          expect(response.data).toEqual(allProductsResponse);
          done();
        }, done);
      });
    });

    describe('get product by id', () => {
      describe('product exists', () => {
        // beforeAll(() => {
        //   provider.addInteraction({
        //     state: 'products id 1 exists',
        //     uponReceiving: 'get product with id 1',
        //     withRequest: {
        //       method: 'GET',
        //       path: '/products/1',
        //     },
        //     willRespondWith: {
        //       status: 200,
        //       headers: {
        //         'Content-Type': 'application/json; charset=utf-8',
        //       },
        //       body: like(productResponse),
        //     },
        //   });
        // });

        it('return product id 1', (done) => {
          productsApi.getById(1).then((response) => {
            expect(response.status).toBe(200);
            expect(response.data).toEqual(productResponse);
            done();
          }, done);
        });
      });

      describe('product does not exists', () => {
        // beforeAll(() => {
        //   provider.addInteraction({
        //     state: 'products id 3 does not exist',
        //     uponReceiving: 'get error product not found',
        //     withRequest: {
        //       method: 'GET',
        //       path: '/products/3',
        //     },
        //     willRespondWith: {
        //       status: 404,
        //       headers: {
        //         'Content-Type': 'application/json; charset=utf-8',
        //       },
        //       body: like(productNotFoundErrorResponse),
        //     },
        //   });
        // });

        it('return not found error', (done) => {
          try {
            expect(productsApi.getById(3)).rejects.toThrow(
              'Request failed with status code 404',
            );
          } finally {
            done();
          }
        });
      });
    });
  },
);
