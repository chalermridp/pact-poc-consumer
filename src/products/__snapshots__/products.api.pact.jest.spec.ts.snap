// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pact between PactPocConsumer and PactPocProvider with 30000 ms timeout for Pact get all products return all products 1`] = `
Object {
  "id": Any<String>,
  "is_active": Any<Boolean>,
  "name_en": Any<String>,
  "price": Any<Number>,
  "sku": Any<String>,
}
`;

exports[`Pact between PactPocConsumer and PactPocProvider with 30000 ms timeout for Pact get all products return all products 2`] = `
Object {
  "id": Any<String>,
  "is_active": Any<Boolean>,
  "name_en": Any<String>,
  "price": Any<Number>,
  "sku": Any<String>,
}
`;

exports[`Pact between PactPocConsumer and PactPocProvider with 30000 ms timeout for Pact get product by id product exists return product id 1 1`] = `
Object {
  "id": Any<String>,
  "is_active": Any<Boolean>,
  "name_en": Any<String>,
  "price": Any<Number>,
  "sku": Any<String>,
}
`;
